internal/jimm/login/login.go

```diff
@@ -0,0 +1,237 @@
+	// into the uri.
+	//
+	// The interval, expiry and device code and used to poll the token endpoint for completion.
+	Device(ctx context.Context) (*oauth2.DeviceAuthResponse, error)
+
+	// DeviceAccessToken continues and collect an access token during the device login flow
+	// and is step TWO.
+	//
+	// See Device(...) godoc for more info pertaining to the flow.
+	DeviceAccessToken(ctx context.Context, res *oauth2.DeviceAuthResponse) (*oauth2.Token, error)
+
+	// ExtractAndVerifyIDToken extracts the id token from the extras claims of an oauth2 token
+	// and performs signature verification of the token.
+	ExtractAndVerifyIDToken(ctx context.Context, oauth2Token *oauth2.Token) (*oidc.IDToken, error)
+
+	// Email retrieves the users email from an id token via the email claim
+	Email(idToken *oidc.IDToken) (string, error)
+
+	// MintSessionToken mints a session token to be used when logging into JIMM
+	// via an access token. The token only contains the user's email for authentication.
+	MintSessionToken(email string) (string, error)
+
+	// VerifySessionToken symmetrically verifies the validty of the signature on the
+	// access token JWT, returning the parsed token.
+	//
+	// The subject of the token contains the user's email and can be used
+	// for user object creation.
+	// If verification fails, return error with code CodeInvalidSessionToken
+	// to indicate to the client to retry login.
+	VerifySessionToken(token string) (jwt.Token, error)
+
+	// UpdateIdentity updates the database with the display name and access token set for the user.
+	// And, if present, a refresh token.
+	UpdateIdentity(ctx context.Context, email string, token *oauth2.Token) error
+
+	// VerifyClientCredentials verifies the provided client ID and client secret.
+	VerifyClientCredentials(ctx context.Context, clientID string, clientSecret string) error
+
+	// AuthenticateBrowserSession updates the session for a browser, additionally
+	// retrieving new access tokens upon expiry. If this cannot be done, the cookie
+	// is deleted and an error is returned.
+	AuthenticateBrowserSession(ctx context.Context, w http.ResponseWriter, req *http.Request) (context.Context, error)
+}
+
+// loginManager provides a means to manage identities within JIMM.
+type loginManager struct {
+	store              *db.Database
+	authSvc            *openfga.OFGAClient
+	oAuthAuthenticator OAuthAuthenticator
+	jimmTag            names.ControllerTag
+}
+
+// NewLoginManager returns a new loginManager that persists the roles in the provided store.
+func NewLoginManager(store *db.Database, authSvc *openfga.OFGAClient, oAuthAuthenticator OAuthAuthenticator, jimmTag names.ControllerTag) (*loginManager, error) {
+	if store == nil {
+		return nil, errors.E("login store cannot be nil")
+	}
+	if authSvc == nil {
+		return nil, errors.E("login authorisation service cannot be nil")
+	}
+	if oAuthAuthenticator == nil {
+		return nil, errors.E("oauth service cannot be nil")
+	}
```

Comment:
nitpick: add
```
if jimmTag.Id() == "" {
   return nil, errors.E("invalid jimm controller tag")
}
```

Reply:
I'll add this in the next PR in the chain to skip re-running tests on this PR.